File: java/Log.java
import java.util.Arrays;
import java.lang.StackTraceElement;

// ===================== CONSTANTS ===================== //

class Log
{
	private static final String COLOUR_ERROR   = Colours.RED;
	private static final String COLOUR_WARNING = Colours.YELLOW;
	private static final String COLOUR_SUCCESS = Colours.GREEN;
	private static final String COLOUR_DEBUG   = Colours.CYAN;
	private static final String COLOUR_INFO    = Colours.BLUE;
	private static final String COLOUR_VERBOSE = Colours.MAGENTA;
	private static final String COLOUR_TRACE   = Colours.BLACK;

	private static final String PREFIX_ERROR   = "X";
	private static final String PREFIX_WARNING = "!";
	private static final String PREFIX_SUCCESS = "~";
	private static final String PREFIX_DEBUG   = "#";
	private static final String PREFIX_INFO    = ">";
	private static final String PREFIX_VERBOSE = "@";

	private static final String [] PREFIXES = {
		PREFIX_ERROR,
		PREFIX_WARNING,
		PREFIX_SUCCESS,
		PREFIX_DEBUG,
		PREFIX_INFO,
		PREFIX_VERBOSE
	};

	private static final String [] COLOURS = {
		COLOUR_ERROR,
		COLOUR_WARNING,
		COLOUR_SUCCESS,
		COLOUR_DEBUG,
		COLOUR_INFO,
		COLOUR_VERBOSE
	};

	public enum Level
	{
		ERROR,
		WARNING,
		SUCCESS,
		DEBUG,
		INFO,
		VERBOSE,

		NUM_LEVELS
	};

	private static final String [] LEVEL_NAMES = {
		"ERROR",
		"WARNING",
		"SUCCESS",
		"DEBUG",
		"INFO",
		"VERBOSE"
	};

	// ===================== FLAGS ===================== //

	private static boolean [] SUPPRESSED = {
		false,
		false,
		false,
		false,
		false,
		false
	};

	private static boolean traceFile = false;
	private static boolean traceLine = false;
	private static boolean traceCaller = false;

	private static boolean enableLogs = true;
	private static boolean enableTrace = false;
	private static boolean enableColour = true;

	// ===================== INTERNAL FUNCTIONS ===================== //

	private static String getTrace()
	{
		if (!enableTrace)
		{
			return "";
		}

		StackTraceElement [] stack = Thread.currentThread().getStackTrace();

		String file = "";
		String line = "";
		String caller = "";

		if (traceFile)
		{
			file = stack[4].getFileName() + ":";
		}

		if (traceLine)
		{
			line = stack[4].getLineNumber() + ":";
		}

		if (traceCaller)
		{
			caller = stack[4].getMethodName() + ":";
		}


		if (enableColour)
		{
			return String.format("%s%s%s%s%s ", COLOUR_TRACE, file, line, caller, Colours.RESET);
		}
		else
		{
			return String.format("%s%s%s ", file, line, caller);
		}

	}

	private static String getPrefix(Level level)
	{
		int index = level.ordinal();

		if (enableColour)
		{
			return String.format("%s%s%s ", COLOURS[index], PREFIXES[index], Colours.RESET);
		}
		else
		{
			return String.format("%s ", PREFIXES[index]);
		}
	}

	private static String formatMessage(Level level, String format, Object... args)
	{
		if (SUPPRESSED[level.ordinal()] || !enableLogs)
		{
			return "";
		}
		else
		{
			return String.format("%s%s%s\r\n", getPrefix(level), getTrace(), String.format(format, args));
		}
	}

	private static void printInvalidLogLevel(Level level)
	{
		System.out.printf("'%s' is an invalid log level\r\n", level);
		System.out.printf("Valid log levels: %s\r\n", Arrays.toString(LEVEL_NAMES));
	}

	// ===================== LOGGING API ===================== //

	public static void enable()
	{
		enableLogs = true;
	}

	public static void disable()
	{
		enableLogs = false;
	}

	public static void suppress(Level level)
	{
		if (level.ordinal() < Level.NUM_LEVELS.ordinal())
		{
			SUPPRESSED[level.ordinal()] = true;
		}
		else
		{
			printInvalidLogLevel(level);
		}
	}

	public static void show(Level level)
	{
		if (level.ordinal() < Level.NUM_LEVELS.ordinal())
		{
			SUPPRESSED[level.ordinal()] = false;
		}
		else
		{
			printInvalidLogLevel(level);
		}
	}

	public static void colourize()
	{
		enableColour = true;
	}

	public static void colourless()
	{
		enableColour = false;
	}

	public static void trace(boolean file, boolean line, boolean caller)
	{
		traceFile = file;
		traceLine = line;
		traceCaller = caller;

		if (file || line || caller)
		{
			enableTrace = true;
		}
		else
		{
			enableTrace = false;
		}
	}

	public static void error(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.ERROR, format, args));
	}

	public static void warning(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.WARNING, format, args));
	}

	public static void success(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.SUCCESS, format, args));
	}

	public static void debug(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.DEBUG, format, args));
	}

	public static void info(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.INFO, format, args));
	}

	public static void verbose(String format, Object ... args)
	{
		System.out.print(formatMessage(Level.VERBOSE, format, args));
	}
}

File: c/log.c

#include <stdio.h>
#include <stdbool.h>
#include <stdarg.h>

#include "colours.h"

#include "log.h"

// ===================== CONSTANTS ===================== //

#define NEWLINE   "\r\n"

#define COLOUR_ERROR     RED
#define COLOUR_WARNING   YELLOW
#define COLOUR_SUCCESS   GREEN
#define COLOUR_DEBUG     CYAN
#define COLOUR_INFO      BLUE
#define COLOUR_VERBOSE   MAGENTA
#define COLOUR_TRACE     BLACK

#define PREFIX_ERROR     "X"
#define PREFIX_WARNING   "!"
#define PREFIX_SUCCESS   "~"
#define PREFIX_DEBUG     "#"
#define PREFIX_INFO      ">"
#define PREFIX_VERBOSE   "@"

static char const * prefixes [NUM_LOG_LEVELS] = {
	PREFIX_ERROR,
	PREFIX_WARNING,
	PREFIX_SUCCESS,
	PREFIX_DEBUG,
	PREFIX_INFO,
	PREFIX_VERBOSE
};

static char const * colours [NUM_LOG_LEVELS] = {
	COLOUR_ERROR,
	COLOUR_WARNING,
	COLOUR_SUCCESS,
	COLOUR_DEBUG,
	COLOUR_INFO,
	COLOUR_VERBOSE
};

// ===================== FLAGS ===================== //

static bool suppressed [NUM_LOG_LEVELS] = {
	false,
	false,
	false,
	false,
	false,
	false
};

static bool trace_file = false;
static bool trace_line = false;
static bool trace_caller = false;

static bool enable = true;
static bool trace = false;
static bool colourize = true;

// ===================== LOGGING API ===================== //

void log_enable(void)
{
	enable = true;
}

void log_disable(void)
{
	enable = false;
}

void log_suppress(Log_Level level)
{
	if (level < NUM_LOG_LEVELS)
	{
		suppressed[level] = true;
	}
}

void log_show(Log_Level level)
{
	if (level < NUM_LOG_LEVELS)
	{
		suppressed[level] = false;
	}
}

void log_colourize(void)
{
	colourize = true;
}

void log_colourless(void)
{
	colourize = false;
}

void log_trace(bool file, bool line, bool caller)
{
	trace_file = file;
	trace_line = line;
	trace_caller = caller;

	trace = file || line || caller;
}

void log_print(char const * file, unsigned int line, char const * caller, Log_Level level, char const * format, ...)
{
	if (format == NULL || caller == NULL || file == NULL || level >= NUM_LOG_LEVELS)
	{
		return;
	}

	if (suppressed[level] || !enable)
	{
		return;
	}

	if (trace)
	{
		if (colourize)
		{
			printf("%s", COLOUR_TRACE);
		}

		if (trace_file)
		{
			printf("%s:", file);
		}

		if (trace_line)
		{
			printf("%d:", line);
		}

		if (trace_caller)
		{
			printf("%s:", caller);
		}

		printf(" ");
	}

	printf("%s%s%s ",
		colourize ? colours[level] : "",
		prefixes[level],
		colourize ? RESET : ""
	);

	va_list args;
	va_start(args, format);
	vprintf(format, args);
	va_end(args);

	printf(NEWLINE);
}
